### PR TITLE
RPE-304: resubmission

### PR DESCRIPTION
**Pull request by doncem**
### JIRA link (if applicable) ###

[Store message id in a unique db column](https://tools.hmcts.net/jira/browse/RPE-304)

### Change description ###

Uniqueness is identified as message ID + state.created. Otherwise rule is waived. No DB migration required

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java [hunk 1]
```diff
@@ -38,13 +40,24 @@ public UUID send(LetterRequest letter, String serviceName) {
 
         log.info("Generated message: id = {} for letter with print queue id = {}", messageId, letter.type);
 
-        byte[] pdf = pdfCreator.create(letter);
-        Letter dbLetter = new Letter(messageId, serviceName, null, letter.type, pdf);
+        Optional<Letter> result = letterRepository.findByMessageId(messageId);
+
+        Letter dbLetter = result.filter(l -> l.getState().equals(LetterState.Created))
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: It would be good if you log that the letter existed and hence returning existing letter id

File: src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java [hunk 2]
```diff
@@ -50,6 +54,42 @@ public void generates_and_saves_pdf() throws IOException {
         document.validate();
     }
 
+    @Test
+    public void generate_and_save_pdf_and_return_same_id_on_resubmit() throws IOException {
```
Comment (nitinprabhu) [inline src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java]: Please add exception test cases

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java [hunk 3]
```diff
@@ -38,13 +40,24 @@ public UUID send(LetterRequest letter, String serviceName) {
 
         log.info("Generated message: id = {} for letter with print queue id = {}", messageId, letter.type);
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: Update this logger now there is no queue id

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java [hunk 4]
```diff
@@ -38,13 +40,24 @@ public UUID send(LetterRequest letter, String serviceName) {
...
 
-        byte[] pdf = pdfCreator.create(letter);
-        Letter dbLetter = new Letter(messageId, serviceName, null, letter.type, pdf);
+        Optional<Letter> result = letterRepository.findByMessageId(messageId);
+
+        Letter dbLetter = result.filter(l -> l.getState().equals(LetterState.Created))
+            .orElse(getNewLetter(letter, messageId, serviceName));
 
-        letterRepository.save(dbLetter);
         return dbLetter.getId();
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: Please log this id if as it might be useful for tracking

**Comment by nitinprabhu** [comment]
Also worth updating the JIRA as it refers to queue stuff

File: src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java (hunk 2 above)
Comment (doncem) [reply src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java]: Exception comes from `SampleData.letter()` while reading the resource. Would be like testing a test

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java [hunk 5]
```diff
@@ -36,13 +38,31 @@ public UUID send(LetterRequest letter, String serviceName) {
 
         final String messageId = generateChecksum(letter);
 
-        log.info("Generated message: id = {} for letter with print queue id = {}", messageId, letter.type);
+        log.info("Generated message: id = {}", messageId);
 
-        byte[] pdf = pdfCreator.create(letter);
-        Letter dbLetter = new Letter(messageId, serviceName, null, letter.type, pdf);
+        Optional<Letter> result = letterRepository.findByMessageIdOrderByCreatedAtDesc(messageId);
```
Comment (banderous) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: Could you instead put a method on the repository to search by both MessageId and state ie:

if (repository.existsByMessageIdAndState(messageId, LetterState.Created)) {
}

Comment (banderous) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: Could you instead put a method on the repository to search by both MessageId and state ie:

if (repository.existsByMessageIdAndState(messageId, LetterState.Created)) {
}

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java [hunk 6]
```diff
@@ -36,13 +38,31 @@ public UUID send(LetterRequest letter, String serviceName) {
...
+    }
+
+    private UUID getNewLetterId(LetterRequest letterRequest, String messageId, String serviceName) {
+        byte[] pdf = pdfCreator.create(letterRequest);
+
+        Letter letter = new Letter(messageId, serviceName, null, letterRequest.type, pdf);
+
+        letterRepository.save(letter);
+
+        UUID letterId = letter.getId();
```
Comment (banderous) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: you might like to combine this with the preceding line: UUID letterId = letterRepository.save(letter).getId();

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java (hunk 5 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: but i need to return the original letter in case there is a letter in `created` state. ought to be the latest hence the ordering by `created_at`

Comment (banderous) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: You could do findByMessageIdAndStateOrderByCreatedAtDesc, then the filtering could be done by the DB.

Comment (banderous) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: Could also use a spring projection to go straight to id as a UUID, but may not be worth it for this simple case - https://docs.spring.io/spring-data/rest/docs/current/reference/html/#projections-excerpts.projections

File: src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java [hunk 7]
```diff
@@ -50,6 +54,42 @@ public void generates_and_saves_pdf() throws IOException {
...
+        assertThat(result.getState()).isEqualByComparingTo(LetterState.Created);
+
+        // then
+        result.setState(LetterState.Uploaded);
+        letterRepository.saveAndFlush(result);
+        UUID id2 = service.send(sampleRequest, SERVICE_NAME);
+
+        // and
+        assertThat(id1).isNotEqualByComparingTo(id2);
+    }
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java]: missing 'when'?

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java [hunk 8]
```diff
@@ -36,13 +38,32 @@ public UUID send(LetterRequest letter, String serviceName) {
 
         final String messageId = generateChecksum(letter);
 
-        log.info("Generated message: id = {} for letter with print queue id = {}", messageId, letter.type);
+        log.info("Generated message: id = {}", messageId);
 
-        byte[] pdf = pdfCreator.create(letter);
-        Letter dbLetter = new Letter(messageId, serviceName, null, letter.type, pdf);
+        Optional<Letter> result = letterRepository.findByMessageIdAndStateOrderByCreatedAtDesc(
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: perhaps `duplicateLetter`?

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java [hunk 9]
```diff
@@ -36,13 +38,32 @@ public UUID send(LetterRequest letter, String serviceName) {
...
+        return result
+            .map(l -> {
+                UUID id = l.getId();
+                log.info("Same message found already created. Returning {} letter", id);
+                return id;
+            })
+            .orElse(getNewLetterId(letter, messageId, serviceName));
+    }
+
+    private UUID getNewLetterId(LetterRequest letterRequest, String messageId, String serviceName) {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: confusing name, this 'get' saves to db

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java [hunk 10]
```diff
@@ -36,13 +38,32 @@ public UUID send(LetterRequest letter, String serviceName) {
...
-        Letter dbLetter = new Letter(messageId, serviceName, null, letter.type, pdf);
+        Optional<Letter> result = letterRepository.findByMessageIdAndStateOrderByCreatedAtDesc(
+            messageId,
+            LetterState.Created
+        );
+
+        return result
+            .map(l -> {
+                UUID id = l.getId();
+                log.info("Same message found already created. Returning {} letter", id);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: is 'Returning 8eb874b2-e975-4006-aa2c-9883fa66b065 letter' the message you want to log?

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java [hunk 11]
```diff
@@ -36,13 +38,32 @@ public UUID send(LetterRequest letter, String serviceName) {
...
+                return id;
+            })
+            .orElse(getNewLetterId(letter, messageId, serviceName));
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: `.orElseGet` for lazy evaluation

File: src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java [hunk 12]
```diff
@@ -50,6 +54,42 @@ public void generates_and_saves_pdf() throws IOException {
...
+        Letter result = letterRepository.findOne(id1);
+
+        // and
+        assertThat(result.getState()).isEqualByComparingTo(LetterState.Created);
+
+        // then
+        UUID id2 = service.send(sampleRequest, SERVICE_NAME);
+
+        // and
+        assertThat(id1).isEqualByComparingTo(id2);
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java]: also check that only one row was insterted (with currently implementation you save a new one anyway)

Comment (doncem) [reply src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java]: is table identification comparison enough?

File: src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java [hunk 13]
```diff
@@ -50,6 +54,42 @@ public void generates_and_saves_pdf() throws IOException {
...
+        Letter result = letterRepository.findOne(id1);
+
+        // and
+        assertThat(result.getState()).isEqualByComparingTo(LetterState.Created);
+
+        // when
+        UUID id2 = service.send(sampleRequest, SERVICE_NAME);
+
+        // then
+        assertThat(id1).isEqualByComparingTo(id2);
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java]: also check that only one row was insterted to db

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java [hunk 14]
```diff
@@ -36,13 +38,32 @@ public UUID send(LetterRequest letter, String serviceName) {
...
+                log.info("Same message found already created. Returning letter id {} instead", id);
+                return id;
+            })
+            .orElseGet(() -> saveNewLetterAndReturnId(letter, messageId, serviceName));
+    }
+
+    private UUID saveNewLetterAndReturnId(LetterRequest letterRequest, String messageId, String serviceName) {
+        byte[] pdf = pdfCreator.create(letterRequest);
+
+        Letter letter = new Letter(messageId, serviceName, null, letterRequest.type, pdf);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: why `null`?

File: src/test/java/uk/gov/hmcts/reform/sendletter/controllers/SendLetterControllerTest.java [hunk 15]
```diff
@@ -85,22 +82,6 @@ public void should_return_connection_exception_when_service_fails_due_to_thread_
...
-        sendLetter(readResource("letter.json"))
-            .andExpect(status().isBadRequest())
-            .andExpect(content().string(
-                containsString("Exception occurred while parsing letter contents")));
-
-
-        verify(authService).authenticate("auth-header-value");
-        verifyNoMoreInteractions(authService);
-
-    }
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/sendletter/controllers/SendLetterControllerTest.java]: is this related?

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java (hunk 14 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: additional data is not needed any more since we generate pdf in advance

File: src/test/java/uk/gov/hmcts/reform/sendletter/controllers/SendLetterControllerTest.java (hunk 15 above)
Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/sendletter/controllers/SendLetterControllerTest.java]: this was related to service bus when we serialised the letter. not needed now

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java (hunk 14 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/LetterService.java]: it is still needed

File: src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java [hunk 16]
```diff
@@ -21,25 +27,39 @@
...
 
     @Autowired
     private LetterRepository letterRepository;
 
+    private final ObjectMapper mapper = new ObjectMapper();
+
     @Before
     public void setUp() {
-        service = new LetterService(letterRepository);
+        service = new LetterService(letterRepository, mapper);
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java]: inline if mapper is not used anywhere else

File: src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java [hunk 17]
```diff
@@ -50,6 +70,48 @@ public void generates_and_saves_pdf() throws IOException {
...
+        // and
+        verify(letterRepository).save(any(Letter.class));
+    }
+
+    @Test
+    public void generate_and_save_pdf_twice_if_previous_letter_has_been_sent_to_print() throws IOException {
+        // given
+        LetterRequest sampleRequest = SampleData.letter();
+        UUID id1 = service.send(sampleRequest, SERVICE_NAME);
+        Letter result = letterRepository.findOne(id1);
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/sendletter/services/LetterServiceTest.java]: a less generic variable name would be nice
